EIPS/eip-6963.md

```diff
@@ -0,0 +1,211 @@
+
+In this proposal, we provide a solution that optimitizes for interoperability of multiple wallet providers and enables fairer competition by lowering the barrier to entry for new wallet providers to improve the user experience on Ethereum.
+
+This is achieved by introducing a set of window events to provide a two-way communication protocol between Ethereum libraries and injected scripts provided by browser extensions thus enabling users to select their wallet of choice.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Provider Info
+
+Each wallet provider will be exposed with the following interface `ProviderInfo` that will be used to display to the user:
+
+```typescript
+/**
+ * Represents the assets needed to display a wallet
+ */
+interface ProviderInfo {
+  walletId: string;
+  uuid: string;
+  name: string;
+  icon: string;
+}
+```
+
+The values in the `ProviderInfo` MUST be used as follows:
+
+- walletId - globally unique identifier of the wallet provider (eg. `io.dopewallet.extension` or `awesomewallet`)
+- uuid - locally unique of the wallet provider (UUIDv4 compliant)
+- name - human-readable name of the wallet provider (e.g. `DopeWalletExtension` or `AwesomeWallet`)
+- icon - uri encoded image (RFC-3986 complaint)
+
+### EVMProvider
+
+The wallet provider will also expose their own [EIP-1193](./eip-1193.md) provider interface in parallel with the provider info provided above in the following interface `EVMProvider`
+
+```typescript
+interface EVMProvider {
+  info: ProviderInfo;
+  ethereum: EIP1193Provider;
+}
+```
+
+The `EIP1193Provider` interface is documented at [EIP-1193](./eip-1193.md) and can be used to override the `window.ethereum` object once the user as explicitly selected it.
+
+### Window Event
+
+Different wallet providers might inject their scripts at different timeframes plus there is no guarantees that the Ethereum library in the web page will be loaded before all injected scripts are present in the page.
+
+To emit events it must use `window.dispatchEvent` and to observe events it must use `window.addEventListener`. There will be two event interfaces:
+
+```typescript
+// Requesting an EVM provider
+interface EVMProviderRequestedEvent extends Event {
+  type: "evmProviderRequested";
+}
+
+// Annoucing an EVM provider
+interface EVMProviderAnnouncedEvent extends CustomEvent {
+  type: "evmProviderAnnounced";
+  detail: EVMProvider;
+}
+```
+
+Therefore both Wallet providers and the Ethereum library must emit events when they initialize.
+
+```typescript
+// Ethereum library initializes
+window.dispatchEvent(new Event("evmProviderRequested"));
+
+// Wallet provider initializes
+window.dispatchEvent(
+  new CustomEvent("evmProviderAnnounced", { detail: provider })
```

Comment:
Thoughts on "namespacing" these event names? Something like `"evm:announceProvider"` or `"eip6963:announceProvider"`.

Verbs in JavaScript event names are also traditionally written in present tense. I think we should match that convention.

Reply:
Curious why we want the namespace? For example, common browser event types (e.g. `"mousedown"`) aren't namespaced, like `"ui-event-20040104:mousedown"`.

`"announceProvider"` is easier to remember and isn't tightly coupled to this specific snapshot in time (e.g. random EIP number). If a namespace is important, seems like `"evm"` would be better suited than `"eip6963"`.

Reply:
cc @pedrouid  @darrylyeo

Reply:
Using `evm` as a prefix IMO is going to create more problems than it solves

We can all agree on EIP-1193 and EIP-6963 specifications

But it's hard to find consensus over terms like Ethereum and EVM

It's ok to use them as descriptions but I would like to avoid them in naming conventions that can cause future issues with versioning or namespacing